### PR TITLE
Bigdataviewer-imageloader dependency

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

**Comment by KateMoreva** [comment]
Current MoBIE was moved to the `bigdataviewer-imageloader` repo. 
**Refactoring is not included.**

File: pom.xml [hunk 1]
```diff
@@ -150,6 +150,11 @@
 			<artifactId>bigdataviewer-vistools</artifactId>
 			<version>${bigdataviewer-vistools.version}</version>
 		</dependency>
+        <dependency>
+            <groupId>mobie</groupId>
+            <artifactId>bigdataviewer-imageloader</artifactId>
+            <version>1.0.1-SNAPSHOT</version>
```
Comment (constantinpape) [inline pom.xml:156]: Should we wait for merging this until https://github.com/mobie/bigdataviewer-imageloader/pull/16 is merged and released? Then we can use the artifact from scijava.maven and don't need the snapshot dependency here.

**Review by constantinpape — COMMENTED** [review]
Looks good overall. I would suggest we wait till https://github.com/mobie/bigdataviewer-imageloader/pull/16 is merged and released (so we can use the dependency from scijava.maven) and I test it locally before merging this one.

> Refactoring is not included.

What do you mean by this?

**Comment by constantinpape** [comment]
@KateMoreva @tischi @K-Meech 
I get an error when I try to build MoBIE (see stacktrace below), both on this branch and on the current master. Has one of you seen this before and knows how to fix it?
```
Error:java: java.io.IOException: Unexpected char: '}' from org.scijava.plugin.Plugin from java.io.BufferedInputStream@4c2c1035
  	at org.scijava.annotations.IndexReader.next(IndexReader.java:163)
  	at org.scijava.annotations.AbstractIndexWriter.merge(AbstractIndexWriter.java:144)
  	at org.scijava.annotations.AbstractIndexWriter.write(AbstractIndexWriter.java:96)
  	at org.scijava.annotations.AnnotationProcessor.process(AnnotationProcessor.java:93)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:802)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:713)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1043)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1184)
  	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
  	at com.sun.tools.javac.main.Main.compile(Main.java:523)
  	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
  	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
  	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:207)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:486)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:338)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:263)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:216)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1328)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1006)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1073)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:967)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:796)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:378)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:178)
  	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:140)
  	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:297)
  	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:130)
  	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:232)
  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  	at java.lang.Thread.run(Thread.java:748)
```

**Comment by constantinpape** [comment]
@tischi I can't test if this branch works right now because of the build error (see comment above). I have this error both on this branch and on master, so I don't think that it is related to the changes here.
Could you maybe try MoBIE with this branch and see if it works for you?

**Comment by constantinpape** [comment]
Ok, this was just some intelliJ issue, reimporting code and dependencies fixed it.

**Comment by constantinpape** [comment]
@KateMoreva I tried to run `OpenRemotePlatynereisBranch`, both using the latest scijava.maven package of `bigdataviewer-imageloader` and the current master. But it fails with the following error in both cases:
```
[ERROR] java.lang.ArrayIndexOutOfBoundsException: -1
	at java.util.ArrayList.elementData(ArrayList.java:424)
	at java.util.ArrayList.get(ArrayList.java:437)
	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.getParentMenu(IJ1Helper.java:999)
	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.create(IJ1Helper.java:958)
	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.access$700(IJ1Helper.java:930)
	at net.imagej.legacy.IJ1Helper.addMenuItems(IJ1Helper.java:915)
	at net.imagej.legacy.LegacyService.initialize(LegacyService.java:508)
	at org.scijava.service.ServiceHelper.createServiceRecursively(ServiceHelper.java:352)
	at org.scijava.service.ServiceHelper.createExactService(ServiceHelper.java:267)
	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:229)
	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:192)
	at org.scijava.service.ServiceHelper.loadServices(ServiceHelper.java:164)
	at org.scijava.Context.<init>(Context.java:285)
	at org.scijava.Context.<init>(Context.java:234)
	at org.scijava.Context.<init>(Context.java:174)
	at org.scijava.Context.<init>(Context.java:160)
	at net.imagej.ImageJ.<init>(ImageJ.java:77)
	at projects.OpenRemotePlatynereisBranch.main(OpenRemotePlatynereisBranch.java:13)
```

**Comment by KateMoreva** [comment]
> @KateMoreva I tried to run `OpenRemotePlatynereisBranch`, both using the latest scijava.maven package of `bigdataviewer-imageloader` and the current master. But it fails with the following error in both cases:
> (and I double checked, this works on master)
> 
> ```
> [ERROR] java.lang.ArrayIndexOutOfBoundsException: -1
> 	at java.util.ArrayList.elementData(ArrayList.java:424)
> 	at java.util.ArrayList.get(ArrayList.java:437)
> 	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.getParentMenu(IJ1Helper.java:999)
> 	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.create(IJ1Helper.java:958)
> 	at net.imagej.legacy.IJ1Helper$IJ1MenuWrapper.access$700(IJ1Helper.java:930)
> 	at net.imagej.legacy.IJ1Helper.addMenuItems(IJ1Helper.java:915)
> 	at net.imagej.legacy.LegacyService.initialize(LegacyService.java:508)
> 	at org.scijava.service.ServiceHelper.createServiceRecursively(ServiceHelper.java:352)
> 	at org.scijava.service.ServiceHelper.createExactService(ServiceHelper.java:267)
> 	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:229)
> 	at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:192)
> 	at org.scijava.service.ServiceHelper.loadServices(ServiceHelper.java:164)
> 	at org.scijava.Context.<init>(Context.java:285)
> 	at org.scijava.Context.<init>(Context.java:234)
> 	at org.scijava.Context.<init>(Context.java:174)
> 	at org.scijava.Context.<init>(Context.java:160)
> 	at net.imagej.ImageJ.<init>(ImageJ.java:77)
> 	at projects.OpenRemotePlatynereisBranch.main(OpenRemotePlatynereisBranch.java:13)
> ```

This error is related to the ImageJ() and should not affect reading files.

**Comment by constantinpape** [comment]
@tischi I checked this now and it seems to all work. But it requires to install the `bigdataviewer-imageloader` repo locally, because it does not work with the version that is currently available via scijava.maven.
Do you want to go ahead and merge this or should we first make a new release of the imageloader repo so that we have a scijava.maven package that works for us?

**Comment by KateMoreva** [comment]
> @tischi I checked this now and it seems to all work. But it requires to install the `bigdataviewer-imageloader` repo locally, because it does not work with the version that is currently available via scijava.maven.
> Do you want to go ahead and merge this or should we first make a new release of the imageloader repo so that we have a scijava.maven package that works for us?

I think a good way could be:

1. Publish a new version of imageloader repo (non SNAPSHOT version)
2. Merge this PR with changes in pom
Then we would have a fully functional MoBIE with the dependency
3. Solve all conflicts with PR in the repo (WIP)
4. Publist the refactored version
5. Update dependency version

**Comment by constantinpape** [comment]
> I think a good way could be:
> 
>     1. Publish a new version of imageloader repo (non SNAPSHOT version)
> 
>     2. Merge this PR with changes in pom
>        Then we would have a fully functional MoBIE with the dependency
> 
>     3. Solve all conflicts with PR in the repo (WIP)
> 
>     4. Publist the refactored version
> 
>     5. Update dependency version

That sounds good! Please go ahead and do this :).

**Comment by KateMoreva** [comment]
@tischi, @K-Meech Should I merge this PR then?

**Comment by constantinpape** [comment]
@KateMoreva I think you still need to make a new release in https://github.com/mobie/bigdataviewer-imageloader before the maven package is updated, or am I missing something?

**Comment by KateMoreva** [comment]
> @KateMoreva I think you still need to make a new release in https://github.com/mobie/bigdataviewer-imageloader before the maven package is updated, or am I missing something?

The release has been made (https://maven.scijava.org/#nexus-search;quick~bigdataviewer-imageloader), the GitHub release is not that important.

**Comment by constantinpape** [comment]
> > 
> 
> The release has been made (https://maven.scijava.org/#nexus-search;quick~bigdataviewer-imageloader), the GitHub release is not that important.

Ok, thanks for clarifying! Then this is good to be merged from my side.

**Comment by K-Meech** [comment]
Yes - looks good! I still think that the files under /projectcreator/n5 (https://github.com/mobie/mobie-viewer-fiji/tree/master/src/main/java/de/embl/cba/mobie/projectcreator/n5)  could also move to this new repo. (They're the writers from imgplus to n5 & ome-zarr). But I can always move these later in a separate PR :)

**Comment by constantinpape** [comment]
> Yes - looks good! I still think that the files under /projectcreator/n5 (https://github.com/mobie/mobie-viewer-fiji/tree/master/src/main/java/de/embl/cba/mobie/projectcreator/n5) could also move to this new repo. (They're the writers from imgplus to n5 & ome-zarr). But I can always move these later in a separate PR :)

Yes, they should def. go into the imageloader repo as well, but let's do that in a separate PR :). I am merging this one now :).

**Pull request merged**